amqVivaceShortcuts.user.js

```diff
@@ -296,12 +301,24 @@ const optimizedShortcuts = (targets) => {
   let minPos = Infinity;
   let bestSubstring = "";
   let shortcuts = [];
+  let altShortcuts = [];
   let currentLength = 0;
 
-  for (const substring of sortedSubstrings) {
+  for (let substring of sortedSubstrings) {
     const newLength = substring.length;
 
     // Search for longer substrings only if there are not enough shortcuts yet, but display *all* the shortest ones
+    if (newLength > currentLength) {
+      let temp = [];
+      for (const altSubString of altShortcuts) {
+        if (altSubString.length == currentLength) {
+          shortcuts.push(altSubString);
+        } else {
+          temp.push(altSubString);
+        }
+      }
+      altShortcuts = temp;
+    }
     if (newLength > currentLength && shortcuts.length >= NUM_SHORTCUTS) break;
 
     const suggestions = getSuggestions(substring);
```

Comment:
```suggestion
    // When searching for longer substrings, first pick those from the altShortcuts list
    if (newLength > currentLength) {
      altShortcuts = altShortcuts.filter((s) => {
        // Move altShortcuts of the currentLength to the shortcuts list
        if (s.length === currentLength) {
          shortcuts.push(s);
          return false;
        }

        // Keep the others in the altShortcuts list
        return true;
      });

      // Re-check the stopping condition after moving the altShortcuts to the shortcuts list
      if (shortcuts.length >= NUM_SHORTCUTS) break;
    }

    const suggestions = getSuggestions(substring);
```

Using `.filter()` makes it more clear that you are removing items from `altShortcuts` (to move them to the `shortcuts` list). Also this way you don't need a `temp` array.
The stopping condition should be checked again after this, since the `shortcuts` length has changed.